add akka-typed project with generic ActorRef

rkuhn opened the pull request:
This is the first step towards more type-safety in Actor interactions,
comprising:
- generic ActorRef[T] that only accepts T messages
- generic ActorSystem[T] extends ActorRef[T](sending to the guardian,
  whose Props[T] are provided for ActorSystem construction)
- removed the Actor trait: everything in there has been made into
  messages and signals
- new Behavior[T] abstraction that consumes messages (of type T) or
  Signals (lifecycle hooks, Terminated, ReceiveTimeout, Failed),
  producing the next Behavior[T] as the result each time
- the ask pattern is provided and yields properly typed Futures
- variants of ActorContext are provided for synchronous testing of
  Behaviors

All of this is implemented without touching code outside akka-typed
(apart from making guardianProps configurable), creating wrapper objects
around ActorRef, ActorContext, ActorSystem, Props and providing an Actor
implementation that just runs a Behavior.

rkuhn commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+class LineNumbers {
+  import LineNumbers._
+
+  // FIXME: this needs memoization with an LRU cache
```
I have a wait-free ConcurrentCache implementation basically ready, but I have not yet benchmarked it.

ben-manes replied:
Alternatively, [ConcurrentLinkedHashMap](https://code.google.com/p/concurrentlinkedhashmap/) is easy to embed. A more feature-rich rewrite is in active [development](https://github.com/ben-manes/caffeine).

rkuhn replied:
I quickly skimmed it and it is a lot of code; I’ll leave this fixme here and settle on an implementation later, thanks for the pointers in any case!

drewhk commented on akka-typed/src/test/scala/akka/typed/PerformanceSpec.scala:
```diff
@@ -0,0 +1,72 @@
+            val stop = Deadline.now
+
+            val rate = 2L * count * pairs * pings / (stop - start).toMillis
+            info(s"messaging rate was $rate/ms")
```
Hm, consider JMH later? Or this is just for a rough validation (like I have with the interpreter)?

rkuhn replied:
Yes, this was just to see that there are no huge problems.

drewhk commented:
Very good (I have not reviewed, just the fact that there is a PR)! I like this direction both in streams and actors that the behavioral models are detached from the execution environment. For example a OneBoundedInterpreter may be embeddable in a typed Behavior, and then executed in some lightweight RS optimized executor.

drewhk commented:
Btw I am also thinking about a verification framework, it would heavily benefit from this separation.

drewhk commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+  final val debug = false
+}
+
+class LineNumbers {
```
I guess this will be useful for quite a lot of stuff. Can't we migrate this class to release-2.3?

(and probably missing an INTERNAL API marker)

ktoso replied:
Oh yeah, we'll definitely want to try using it in streams - perhaps "that merge you defined on line 23 is wrong" is doable with this? I'll have to read more into it, very awesome indeed :-)

rkuhn replied:
Before we do such a thing I’d like to pose the question whether we really want to maintain a byte-code parsing library ;-) (my vote is obviously positive, but there might be things I’m missing). And I agree with the applicability in other places, including every Java lambda ever written—although I have not tested that yet.

drewhk replied:
I vote +1. We can extract it from Akka even, that way we might get more external help given it is quite small. I am willing to rewrite it in Java even if it is considered necessary.

drewhk commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,399 @@
+   * exclusively. When both sub-behaviors respond to a [[Failed]] signal, the
+   * response with the higher precedence is chosen (see [[Failed$]]).
+   */
+  case class Or[T](left: Behavior[T], right: Behavior[T]) extends Behavior[T] {
```
Would be nice to have a SpitThenJoin, where it starts working as an Or, and when individual behaviors become somethings like Behavior.Join, it stops becoming an Or, and transitions to a next top-level behavior.

drewhk replied:
Of course the same can be done for And

rkuhn replied:
I was thinking about adding an `Escape(nextBehavior)` return that either side could use to shed the `Or`, but a `Join` makes sense as well. This is precisely the reason why this is not yet included :-)

drewhk commented on akka-typed/src/main/scala/akka/typed/ActorRef.scala:
```diff
@@ -0,0 +1,75 @@
+ * [[akka.actor.ActorSystem!.eventStream EventStream]] on a best effort basis
+ * (i.e. this delivery is not reliable).
+ */
+abstract class ActorRef[-T] extends java.lang.Comparable[ActorRef[_]] { this: ScalaActorRef[T] ⇒
```
How does it work out when you have to use both the normal ActorRef (old code) and this new typed one? I mean the names are the same (I know, you can rename on import, but still).

rkuhn replied:
Good point: for Java that will look awkward. Ideas, anyone?

vkostyukov replied:
`TypedActorRef`?

hepin1989 replied:
it has been laid in the ```typed``` package

rkuhn replied:
We’ll do some further brain storming and until a good proposal has been found I’ll leave it like this.

@hepin1989 In Java you cannot `import akka.typed`, so mixing typed and untyped ActorRef usage in the same file will need explicit qualification for one of them in all locations, which is quite verbose.

hepin1989 replied:
phantom types?

rkuhn replied:
What do you mean? (those phantom types I know are not available in Java, and I would not see how they would relate to the problem)

hepin1989 replied:
I think that would help to combine two ActorRef as one.sorry for the noise.

drexin replied:
How about naming this class `TypedActorRef` and having an alias to `ActorRef` in the package object? This way we have the nice naming in Scala and a working solution for Java.

viktorklang replied:
Would there be a risk of confusion with `TypedActor`?

drexin replied:
Good point. I think it's about as easy to confuse as `akka.actor.ActorRef` and `akka.typed.ActorRef`, maybe even less, because the name is slightly different. But as there is no relationship in the types, the compiler will tell you, as opposed to e.g. `scala.collection.Seq` and `mutable` vs `immutable`. ;-)

viktorklang replied:
Fair enough :)

rkuhn replied:
While writing the docs the flow of the text (and my general way of talking about things) suggested that the two main concepts in akka-typed are “behaviors” and “addresses”. Therefore I’m currently leaning towards renaming ActorRef to ActorAddress, which would both be more consistent and avoid this problem. Thoughts?

drewhk commented on akka-typed/src/main/scala/akka/typed/Effects.scala:
```diff
@@ -0,0 +1,106 @@
+  import akka.{ actor ⇒ a }
+  import Effect._
+
+  private val eq = new ConcurrentLinkedQueue[Effect]
```
well, maybe a longer, but more descriptive name?

rkuhn replied:
yup.

huntc commented:
Would it be possible to provide an example where there is an actor that receives two messages?

rkuhn commented:
@huntc in all the tests actors receive a multitude of commands—and where that is not enough you can `ctx.createWrapper(<some function>)` that wraps any other message type into something that is a subtype of the current Actor’s input type. See for example ActorContextSpec.

huntc commented:
> @huntc in all the tests actors receive a multitude of commands—and where that is not enough you can ctx.createWrapper(<some function>) that wraps any other message type into something that is a subtype of the current Actor’s input type. See for example ActorContextSpec.

I guess that I felt that the examples were a bit buried. It would be nice to have something represented in this PR's description instead of having to wade through things. I shall now do just that though.

huntc commented on akka-typed/src/test/scala/akka/typed/ActorContextSpec.scala:
```diff
@@ -0,0 +1,508 @@
+  case class BecomeInert(replyTo: ActorRef[BecameInert.type]) extends Command
+  case object BecameInert extends Event
+
+  def subject(monitor: ActorRef[GotSignal]): Behavior[Command] =
```
I could certainly be missing something, but it seems then that this function is effectively the new way to declare an Actor... an Actor being a combination of its ActorContext and Behaviour. Does that then mean that an Actor's behaviour must be a type hierarchy related to some parent type, in this case `Command`?

drewhk replied:
No, this Command trait at L12 is scoped to this Spec, these are just test classes.

huntc commented on akka-typed/src/test/scala/akka/typed/ActorContextSpec.scala:
```diff
@@ -0,0 +1,508 @@
+        case SetTimeout(d, replyTo) ⇒
+          ctx.setReceiveTimeout(d)
+          replyTo ! TimeoutSet
+          Same
```
Is it possible to somehow conceive that the handling of a message yields both a new actor context and a behaviour? It seems inpure to mutate the actor context.

drewhk replied:
Well, the context *is* impure. And it does not make sense to change it, because the context is the execution engine, for example an underlying ActorCell that is already wired to a supervision hierarchy. Imagine changing suddenly from an actor backed context to a test-context runtime. But maybe I wrong and then @rkuhn will correct me :)

rkuhn replied:
Endre is right in that the context necessarily is impure, but not right in that it does not make sense to expose Actors that are exactly as pure as they can be :-) The Pure class that I have over in my development branch is not part of this PR because it is not ready yet. The idea is that invoking the PureBehavior will return effects to be performed and the next behavior—or more precisely the execution of the monadically represented effects results in the next behavior. It is obvious and unavoidable that this will have severe performance implications, therefore it cannot be the basic model that we expose.

drewhk replied:
Ah, but that is slightly different. I thought that @huntc wanted to return a new *context*. Of course returning *effects* is fine, for example returning message sends instead of calling ! makes it possible to basically intercept every possible actor event.

huntc replied:
Nice one. I like the idea of having the pure and impure approaches. I would tend then to go pure until my profiling suggests otherwise. :-)


> On 18 Jan 2015, at 5:30 pm, Roland Kuhn <notifications@github.com> wrote:
> 
> In akka-typed/src/test/scala/akka/typed/ActorContextSpec.scala:
> 
> > +        case Renew(replyTo) ⇒
> > +          replyTo ! Renewed
> > +          subject(monitor)
> > +        case Throw(ex) ⇒
> > +          throw ex
> > +        case MkChild(name, mon, replyTo) ⇒
> > +          val child = name match {
> > +            case None    ⇒ ctx.spawn(Props(subject(mon)))
> > +            case Some(n) ⇒ ctx.spawn(Props(subject(mon)), n)
> > +          }
> > +          replyTo ! Created(child)
> > +          Same
> > +        case SetTimeout(d, replyTo) ⇒
> > +          ctx.setReceiveTimeout(d)
> > +          replyTo ! TimeoutSet
> > +          Same
> Endre is right in that the context necessarily is impure, but not right in that it does not make sense to expose Actors that are exactly as pure as they can be :-) The Pure class that I have over in my development branch is not part of this PR because it is not ready yet. The idea is that invoking the PureBehavior will return effects to be performed and the next behavior—or more precisely the execution of the monadically represented effects results in the next behavior. It is obvious and unavoidable that this will have severe performance implications, therefore it cannot be the basic model that we expose.
> 
> —
> Reply to this email directly or view it on GitHub.
>

viktorklang commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -181,6 +183,9 @@ private[akka] final class PromiseActorRef private (val provider: ActorRefProvide
   import AbstractPromiseActorRef.watchedByOffset
 
+  // This is necessary for weaving the PromiseActorRef into the asked message, i.e. the replyTo pattern.
+  @volatile var messageName = "unknown"
```
It doesn't say why it isn't passed into the constructor?

rkuhn replied:
Hen, meet egg. Egg, meet hen.

viktorklang replied:
What's the C&E situation here?

"+        val a = PromiseActorRef(ref.provider, timeout, targetName = actorSel, sender)
+        a.messageName = message.getClass.getName"

viktorklang commented on akka-actor/src/main/scala/akka/pattern/AskSupport.scala:
```diff
@@ -330,15 +335,14 @@ private[akka] object PromiseActorRef {
     val f = scheduler.scheduleOnce(timeout.duration) {
       result tryComplete Failure(
-        new AskTimeoutException(s"""Ask timed out on [$targetName] after [${timeout.duration.toMillis} ms]. Sender[$sender] sent message of type "$messageClassName"."""))
+        new AskTimeoutException(s"""Ask timed out on [$targetName] after [${timeout.duration.toMillis} ms]. Sender[$sender] sent message of type "${a.messageName}"."""))
```
`messageName` is less accurate than `messageClassName` tho?

rkuhn replied:
true, will fix

viktorklang commented on akka-remote/src/main/scala/akka/remote/transport/ThrottlerTransportAdapter.scala:
```diff
@@ -298,7 +298,8 @@ private[transport] class ThrottlerManager(wrappedTransport: Transport) extends A
       val internalTarget = target.asInstanceOf[InternalActorRef]
-      val ref = PromiseActorRef(internalTarget.provider, timeout, target, mode)
+      val ref = PromiseActorRef(internalTarget.provider, timeout, target)
+      ref.messageName = mode.getClass.getName
```
No C&E here either?

rkuhn replied:
Ah, now I hear you.

viktorklang replied:
Chicken & Ear problem?

viktorklang replied:
Sorry! That should be "Chicken IN Ear problem" :)

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+  case class SourceFile(filename: String) extends Result {
+    override def toString = filename
+  }
+  case class SourceFileLines(filename: String, from: Int, to: Int) extends Result {
```
final case class, unless there is a reason

rkuhn replied:
yup

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+        case Some((from, to)) ⇒ SourceFileLines(source.get, from, to)
+      }
+
+    } finally dis.close()
```
can `close` throw?

rkuhn replied:
I guess it can, but why would that be a problem?

drewhk replied:
do we want that exception to bubble up?

rkuhn replied:
meanwhile I tried it and changed it to swallowing any non-InterruptedException

patriknw commented on akka-testkit/src/main/scala/akka/testkit/TestEventListener.scala:
```diff
@@ -86,11 +86,20 @@ abstract class EventFilter(occurrences: Int) {
   def awaitDone(max: Duration): Boolean = {
-    if (todo != Int.MaxValue && todo > 0) TestKit.awaitCond(todo == 0, max, noThrow = true)
+    if (todo != Int.MaxValue && todo > 0) TestKit.awaitCond(todo <= 0, max, noThrow = true)
     todo == Int.MaxValue || todo == 0
```
should this still be `todo == 0` ?

patriknw replied:
yes, it should

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,247 @@
+import scala.collection.immutable.TreeSet
+import scala.collection.immutable.TreeMap
+import akka.util.Helpers
+import akka.{ actor ⇒ a }
```
perhaps use `untyped` as alias?

viktorklang replied:
:+1:

rkuhn replied:
done

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,247 @@
+   * by invoking [[akka.actor.Cancellable!.cancel* cancel]] on the returned
+   * handle.
+   */
+  def schedule[U](delay: FiniteDuration, target: ActorRef[U], msg: U): a.Cancellable
```
I'm not convinced that we should expose this here. It is to easy to get scheduling wrong. Instead we should have a trait for mixing in named timers.

rkuhn replied:
Named timers will be a Behavior decorator (there are no traits to be mixed in anymore), but this feature of delayed message sending is a fundamental one.

drewhk replied:
why is it a fundamental one?

rkuhn replied:
Actors react to two things: messages and timeouts. Since they can only receive messages, timeouts need to be messages as well.

There is an inherent unfixable race between cancellation and expiry, leading to messages being “canceled” while they are in the mailbox already; if someone has a smart idea how we can safely and with low overhead make those invisible to the behavior I’m all ears, but even without this fix I do think that delayed messages are a fundamental operation in Actor systems.

drewhk replied:
>  I do think that delayed messages are a fundamental operation in Actor systems.

It is I guess just a philosophical difference. Either delayed send is the fundamental building block, where timers are just a derived service, or timers are a fundamental building block, where messages representing the timer trigger is the derived service.

> There is an inherent unfixable race between cancellation and expiry, leading to messages being “canceled” while they are in the mailbox already; if someone has a smart idea how we can safely and with low overhead make those invisible to the behavior I’m all ears, but even without this fix I do think that delayed messages are a fundamental operation in Actor systems.

In the discrete-event simulators I've used or implemented there is a similar issue, since removing an enqueued future event (the timer) is costly. Ususal approaches is that either cancel just puts a filter in place in the simulated event-driven element, basically ignoring the timers that match the filter, or cancellable retains a reference to the message and sets a flag when cancelled. The latter would need a wrapper message and a thread-safe flag in the actor case. Not saying it is a super smart method, just wanted to share my experience.

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,247 @@
+ * effects an Actor can perform and replaces created child Actors by
+ * [[Inbox$.sync* a synchronous Inbox]].
+ */
+class DummyActorContext[T](
```
should this be moved to a testkit?

rkuhn replied:
so far I don’t think so: Behavior is constructed such that it can be run in different contexts, so providing those contexts seems natural

patriknw replied:
then it should perhaps be renamed, dummy doesn't sound so serious

rkuhn replied:
right, will do

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,60 @@
+
+  val untyped: ExtendedActorSystem
+
+  lazy val ref = untyped.provider.guardian
```
explicit return type?

rkuhn replied:
yes

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,60 @@
+object ActorSystem {
+  private class Impl[T](_name: String, _config: Config, _cl: ClassLoader, _ec: Option[ExecutionContext], _p: Props[T])
+    extends ActorSystem[T](_name) with ScalaActorRef[T] {
+    val untyped: ExtendedActorSystem = new ActorSystemImpl(_name, _config, _cl, _ec, Some(Props.untyped(_p))).start()
```
override annotation

rkuhn replied:
yup

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,60 @@
+
+  private class Wrapper(val untyped: ExtendedActorSystem) extends ActorSystem[Nothing](untyped.name) with ScalaActorRef[Nothing]
+
+  def apply[T](name: String, guardianProps: Props[T],
```
java api create methods?

rkuhn replied:
Java API has not yet been done, will be the next PR.

patriknw commented on akka-typed/src/main/scala/akka/typed/Ask.scala:
```diff
@@ -0,0 +1,71 @@
+  private[typed] def ask[T, U](actorRef: ActorRef[T], timeout: Timeout, f: ActorRef[U] ⇒ T): Future[U] = {
+    val p = PromiseRef[U](actorRef)(timeout)
+    val m = f(p.ref)
+    p.promiseRef.messageName = m.getClass.getName
```
Is this why you need mutable `promiseRef.messageName`.
perhaps consider providing two constructors, to minimise the the damage?

rkuhn replied:
yes, this has been fixed

patriknw commented on akka-typed/src/main/scala/akka/typed/Behavior.scala:
```diff
@@ -0,0 +1,242 @@
+ */
+abstract class Behavior[T] { self ⇒
+  def management(ctx: ActorContext[T], msg: Signal): Behavior[T]
+  def message(ctx: ActorContext[T], msg: T): Behavior[T]
```
docs?

rkuhn replied:
doen

patriknw commented on akka-typed/src/main/scala/akka/typed/Behavior.scala:
```diff
@@ -0,0 +1,242 @@
+ * Lifecycle signal that is fired upon creation of the Actor. This will be the
+ * first message that the actor processes.
+ */
+final case object PreStart extends Signal
```
should I comment on missing java api, or is that pending?

patriknw commented on akka-typed/src/main/scala/akka/typed/Behavior.scala:
```diff
@@ -0,0 +1,242 @@
+ * creating a derived class or by employing factory methods like
+ * [[Behavior.Simple]], [[Behavior.Static]], [[Behavior.Full]] etc.
+ */
+abstract class Behavior[T] { self ⇒
```
what is the purpose of `self`?

rkuhn replied:
it is needed in `widen` below

patriknw replied:
what is the difference between that and using `this`?

rkuhn replied:
In the code below `this` refers to the anonymous Behavior just created. An alternative to providing the outer Behavior’s reference with the `self` name would typically be to say `Behavior.this` instead of using self-type syntax, but in this case that might not even work. In any case, the need for this went away in my branch since I moved `.widen` out into the ScalaDSL where it belongs, will push soon).

patriknw replied:
ok, thanks for clarifying

patriknw commented on akka-typed/src/main/scala/akka/typed/Behavior.scala:
```diff
@@ -0,0 +1,242 @@
+ * done then the default behavior is to escalate the failure, which amounts to
+ * failing this actor with the same exception that the child actor failed with.
+ */
+final case class Failed(cause: Throwable, child: ActorRef[Nothing]) extends Signal
```
how is this related to ctx.setFailureResponse?

rkuhn replied:
Ah, good catch, forgot to update this ScalaDoc.

aloiscochard commented:
Yay for type-safety! :-)

patriknw commented on akka-typed/src/main/scala/akka/typed/Impl.scala:
```diff
@@ -0,0 +1,101 @@
+  override def preStart(): Unit =
+    next(behavior.management(ctx, PreStart))
+  override def preRestart(reason: Throwable, message: Option[Any]): Unit =
+    next(behavior.management(ctx, PreRestart(reason)))
```
is the akka.actor.Actor default `preRestart` implemented elsewhere or not at all?

rkuhn replied:
It is part of the Behavior (see Full as an example).

patriknw commented on akka-typed/src/main/scala/akka/typed/Props.scala:
```diff
@@ -0,0 +1,55 @@
+   * Props for a Behavior that just ignores all messages.
+   */
+  def empty[T]: Props[T] = EMPTY.asInstanceOf[Props[T]]
+  private val EMPTY: Props[Any] = Props(ScalaDSL.Static[Any] { case _ ⇒ ScalaDSL.Unhandled })
```
`EMPTY` is not scala style

rkuhn replied:
okay

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,399 @@
+      case _ ⇒ Unhandled
+    }
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] = behavior(msg)
+    override def toString = s"Simple(${LN.forClass(behavior.getClass)})"
```
Simple vs Total?

rkuhn replied:
good catch!

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,399 @@
+   * actors themselves.
+   */
+  case class Static[T](behavior: T ⇒ Unit) extends Behavior[T] {
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] = msg match {
```
why `match`?

rkuhn replied:
no clue ;-)

patriknw commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,399 @@
+
+    @tailrec private def run(ctx: ActorContext[T], next: Behavior[T]): Behavior[T] = {
+      setBehavior(ctx, next)
+      if (inbox.hasMessages) run(ctx, behavior.message(ctx, inbox.receiveMsg()))
```
do we need to protect agains infinite loops?

rkuhn replied:
very good question

drewhk replied:
hmm. It just does not feel right. Why would we need to pass messages between the And and Or components? I would rather join them instead at that point to a non-composite behavior. Maybe you have a use case in mind?

rkuhn replied:
I found the need for this while implementing the Receiver: it speaks a query protocol on the one side and receives messages of type T on the other (erased to Any). The two halves are combined with Or in order to keep the types straight, but elements received by the RHS need to go to the LHS in order to be available for querying. Joining is not the right operation in this case.

In a hierarchical state machine I take it you can also send events between the parts, or not?

This decorator is also useful for other resequencing, like sending a message to self during a Failed signal so that common operation logic can be reused (for starting a new actor or whatever).

drewhk replied:
> In a hierarchical state machine I take it you can also send events between the parts, or not?

Umm, not the ones I used. The two parts don't communicate after the split, the way to communicate is to wait for a synchronization event (join -- basically works like a dumb CSP with "blocking" exchange()), or one side can forcefully move the parent state to the next one.

> This decorator is also useful for other resequencing, like sending a message to self during a Failed signal so that common operation logic can be reused (for starting a new actor or whatever).

Hm. I cannot imagine the code right now, maybe true. Still the infinite cycle part feels scary, can't these expressed in some other way?

viktorklang commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+        1 -> Some(d.readUTF())
+      case 3 ⇒ // Integer
+        d.readInt()
+        1 -> None
```
Memoize these?

rkuhn replied:
good idea!

viktorklang commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  private def readConstant(d: DataInputStream, c: Constants): (Int, Option[String]) = {
+    d.readByte() match {
```
`@switch`?

rkuhn replied:
yup

viktorklang commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,264 @@
+      case 1 ⇒ // Utf8
+        1 -> Some(d.readUTF())
+      case 3 ⇒ // Integer
+        d.readInt()
```
May want to use `skipBytes` iso read-n-drop

rkuhn replied:
yup, replaced and collapsed quite some pointless reads

viktorklang commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,247 @@
+   * The list of child Actors created by this Actor during its lifetime that
+   * are still alive, in no particular order.
+   */
+  def children: Iterable[ActorRef[Nothing]]
```
Is this useful given that you wouldn't be able to know what you can send to it?

rkuhn replied:
Yes: you can watch and stop.

jboner replied:
Good point. Should make that clear.

viktorklang commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,247 @@
+   * inform the execution mechanism of the supervisor decision. If this is not
+   * done then the implied decision is to escalate the failure.
+   */
+  def setFailureResponse(decision: Failed.Decision): Unit
```
This sounds like a smell. Is this the best solution?

rkuhn replied:
I tried many different solutions to this problem and side-effecting is unavoidable. Patrik pointed out that this method should be on the `Failed` object, though, which is indeed a nice improvement.

patriknw commented on akka-actor/src/main/scala/akka/actor/ActorSystem.scala:
```diff
@@ -140,7 +140,7 @@ object ActorSystem {
     val cl = classLoader.getOrElse(findClassLoader())
     val appConfig = config.getOrElse(ConfigFactory.load(cl))
-    new ActorSystemImpl(name, appConfig, cl, defaultExecutionContext).start()
+    new ActorSystemImpl(name, appConfig, cl, defaultExecutionContext, None).start()
```
@rkuhn did you think anything more about `system.actorOf` when there is a user provided guardian?

rkuhn replied:
So far I have removed access to the untyped system from `typed.ActorSystem`, I still need to add the runtime check before I can consider this concern fully addressed.

patriknw replied:
right, there are other things that are interesting from the untyped system, but perhaps you have exposed them explicitly in `typed.ActorSystem`. I'll review the change.

rkuhn commented:
Pushed a commit that should address all comments; it also contains an addition for LineNumbers that makes it work on Java 8 lambdas—but only if they are Serializable (otherwise I have found no way to get at the bytecode and Brian Goetz has confirmed that this is intentional).

rkuhn commented:
Sorry, something went wrong (didn’t push all the commits, dunno why).

rkuhn commented:
Now with docs; will squash and rebase after review

ktoso commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -0,0 +1,206 @@
+      extends Command
+    //#chatroom-protocol
+    //#chatroom-behavior
+    private final case class PostSessionMessage(screenName: String, message: String)
```
minor: maybe it's just me, but I'd prefer to have all message definitions together in the 1st block. (seeing a `sealed class` being "afterwards" in a different place seemed weird). Then the 2nd block just being the actor.

rkuhn replied:
but then I’d have to explain the PostSessionMessage at a place where I don’t want to talk about it yet: it is not part of the external protocol

ktoso replied:
fair enough :)

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,238 @@
+   * protocols can be ingested by this Actor. You are strongly advised to cache
+   * these ActorRefs or to stop them when no longer needed.
+   */
+  def createWrapper[U](f: U ⇒ T): ActorRef[U]
```
I wouldn't agree that this is a wrapper - it's more like an adapter. "I want to adapt these incoming U, to the protocol I understand: T". 

Case in point: In the chat room example it doesn't wrap the incoming msg, it *unpacks* and adapts it to fit the protocol this actor understands:

```scala
        val wrapper = ctx.createWrapper {
          p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
        }
```

rkuhn replied:
Good point, this method is one of the oldest parts of the currently remaining code.

rkuhn replied:
(meaning: will change to `createAdapter`)

ktoso replied:
Related question. these are currently child actors AFAICS. Seems in other places the verb `spawn` is used when actors are created. Would it make sense for this to be `spawnAdapter` or is there a reason to use `create`?

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,93 @@
+  private class Wrapper(val untyped: ExtendedActorSystem) extends ActorSystem[Nothing](untyped.name) with ScalaActorRef[Nothing]
+
+  def apply[T](name: String, guardianProps: Props[T],
+               config: Option[Config] = None,
```
So the user *always* defines the root actor now? Interesting.

patriknw commented on akka-actor-tests/src/test/java/akka/event/LoggingAdapterTest.java:
```diff
@@ -128,7 +128,7 @@ protected Void match(Object event) {
                     assertEquals(level, log.level());
-                    assertEquals(mdc, log.mdc().toString());
+                    // assertEquals(mdc, log.mdc().toString());
                     if(cause != null) {
```
really? is this fixed somewhere else?

rkuhn replied:
ouch, this was not supposed to be here, please disregard (rebasing onto master will remove the need for this IDE workaround)

patriknw commented on akka-actor-tests/src/test/java/akka/util/LineNumberSpecCodeForJava.java:
```diff
@@ -0,0 +1,40 @@
+	// System.out.println(s);
+	// return s;
+	// };
+	// }
```
a comment about why this is commented out?

rkuhn replied:
the comment is in LineNumberSpec, but I will add one here as well

patriknw commented on project/AkkaBuild.scala:
```diff
@@ -86,6 +86,12 @@ object AkkaBuild extends Build {
   )
 
+  lazy val typed = Project(
+    id = "akka-typed",
```
should end with `-experimental`

rkuhn replied:
yes, it should

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,93 @@
+  /**
+   * Access to the underlying (untyped) ActorSystem.
+   *
+   * FIXME: must prevent system.actorOf from untyped actors for custom guardians
```
done?

rkuhn replied:
indeed

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,93 @@
+   *
+   * FIXME: must prevent system.actorOf from untyped actors for custom guardians
+   */
+  protected val untyped: ExtendedActorSystem
```
I think there are more things from the untyped actor system that should be accessible:

* name
* settings
* logConfiguration
* startTime
* uptime
* scheduler (but the api of the Scheduler is untyped)
* dispatchers
* mailboxes ?
* isTerminated
* awaitTermination
* registerOnTermination
* registerExtension
* extension
* hasExtension
* provider (mostly for getDefaultAddress)

Feels like we need an `ActorSystemLike` trait?

rkuhn replied:
My plan was to not add things that are not used in our code base yet, then we can gradually move forward. Termination is already covered (I didn’t include the deprecated variants), scheduling is I think covered via ActorContext. I added #16738 concerning Extensions because that will require a bit more thought.

patriknw replied:
periodic scheduling is not covered (in my opinion)
whenTerminated is not the same as registerOnTermination
settings
dispatchers (e.g. for looking up configured EC)

rkuhn replied:
opened #16742 (periodic scheduling) and #16743 (registerOnTermination) and added the rest

patriknw commented on akka-docs/rst/scala/typed-actors.rst:
```diff
@@ -1,6 +1,13 @@
+.. note::
+
+  This module is deprecated as it will be superseded by the :ref:`typed-scala`
+  project which is currently being developed in open preview mode.
```
+1
then there should be some deprecated annotations in akka.actor.TypedActor and a few words in the migration guide

rkuhn replied:
I wonder whether a deprecated annotation is already warranted: for someone whose build does not tolerate warnings it would mean that they need to migrate to akka-typed already. My intention is to add this note here as an advanced notice that TypedActors are going away, but we will probably on deprecate them in `(2.4).next`. This should probably be explained in the migration guide, good point.

patriknw replied:
the I think it should be worded "This module will be deprecated"

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,320 @@
+  }
+
+  // FIXME: this needs memoization with an LRU cache
+  def apply(obj: AnyRef): Result = forObject(obj)
```
is this good enough as java api?

rkuhn replied:
no, it isn’t, will fix

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,320 @@
+      skipInterfaceInfo(dis)
+      skipFields(dis)
+      val lines = readMethods(dis, filter)
+      val source = readAttributes(dis)
```
are we sure that readMethods and readAttributes will not throw anything because of "unexpected" class format?

rkuhn replied:
my original plan was to let that bubble up, but it should probably turn such things into NoSourceInfo, you’re right

patriknw replied:
log it, but continue with NoSourceInfo feels safer in this case, since we can't test on all platforms

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,320 @@
+    }
+  }
+
+  private def readMethod(d: DataInputStream, code: Int, lnt: Int, filter: Option[String])(implicit c: Constants): Option[(Int, Int)] = {
```
Is `lnt` a good variable name? (also very close to `Int` for some fonts)

rkuhn replied:
no, it is not, fixed

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,320 @@
+  private def skip(d: DataInputStream, length: Int): Unit =
+    if (d.skipBytes(length) != length) throw new IllegalArgumentException("class file ends prematurely")
+
+  private def dbg(s: String): Unit = if (debug) println("LNB: " + s)
```
note that the `if (debug)` must be moved to the caller to avoid the String allocations

rkuhn replied:
I verified that this method is emitted empty when `debug == false`.

patriknw replied:
so you mean that the string instance s"  tag ${c(tag)} ($length bytes)" is never created on the call side?

rkuhn replied:
ah, right; it probably will be elided by HotSpot, I guess, but better be sure

patriknw replied:
I doubt it can. It probably doesn't know if the involved `toString` methods have other side effects.

patriknw commented on akka-actor/src/main/scala/akka/util/LineNumbers.scala:
```diff
@@ -0,0 +1,320 @@
+          }
+        }
+      dbg(s"  yielded $attributes")
+      attributes.flatten.headOption
```
is it worth hand-rolling these things to while loops?

rkuhn replied:
not at this point; the LRU cache shall later make sure that this is not often called

patriknw replied:
ok

patriknw commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -0,0 +1,206 @@
+
+    final case class CurrentCount(x: Int)
+
+    val counter: Behavior[Command] = behavior(0)
```
`val` in `object` is not something we want to show

rkuhn replied:
why?

patriknw replied:
we should not encourage globals (and the `behavior(0)` instance will never be garbage collected)

rkuhn replied:
WDYM globals? We have many vals in objects, they are immutable values after all. The beauty of this model is that I can start 1000000 counters and they will all share the exact same object for their behavior (until they start counting, but for a stateless Actor this pattern could save a lot of heap space—we could even cache the first 100 counter behaviors if we intend to have LOTS of them running).

patriknw replied:
yes, it works for immutable and it would be great to clarify that in the docs if that is not already done.
I just don't want users to think that this is always the recommended way to do it, then we are back before akka 2.0. Users tend to copy without thinking.

patriknw commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -0,0 +1,206 @@
+  def `must say hello`(): Unit = {
+    //#hello-world
+    import HelloWorld._
+    import scala.concurrent.ExecutionContext.Implicits.global
```
why should we encourage usage of that?

rkuhn replied:
What is the alternative? We cannot use the ActorSystem’s dispatcher since we want to run callbacks after that is shut down.

patriknw replied:
um, makes sense for the terminate future
I would make that explicit, and also not use a `for` (what happens if the ask fails?)

rkuhn replied:
good points, will fix

patriknw commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -0,0 +1,206 @@
+        Stopped
+      }
+
+    val system = ActorSystem("counterBecome", Props(Full[CurrentCount] {
```
when is this shutdown?

rkuhn replied:
I forgot to delete this test case; the chat room proved to be a better example for the docs.

patriknw commented on akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala:
```diff
@@ -0,0 +1,206 @@
+            val wrapper = ctx.spawnAdapter {
+              p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
+            }
+            client ! SessionGranted(wrapper)
```
let's say that we didn't care about the screenName and only wanted to reply back with a `ctx.self` narrowed as a `ActorRef[PostMessage]`.  how is that done? should that be mentioned in the doc describing this sample?

rkuhn replied:
That just works (ActorRef is contravariant in T), but you’re right that that should be mentioned somewhere in the text. Will think about the best placement.

rkuhn replied:
Ah, misread slightly: you meant `ActorRef[PostSessionMessage]`, right?

patriknw replied:
no I meant `PostMessage`, but then it would also be a `Command`

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,238 @@
+   * processing its current message. Nothing happens if no Actor exists with the
+   * given name.
+   */
+  def stop(childName: String): Unit
```
thinking if using only the name is a good thing
* sometimes you might want to stop a specific incarnation (and not accidentally the one that happens to be there at the moment)
* kind of strange to use a name for an anonymous child

rkuhn replied:
Accepting an ActorRef does not convey the information that only direct child actors are permissible while giving a name does not verify the incarnation ID. What a pickle. We could of course completely refactor all of Akka to use path elements of type `(String, Int)` instead of just `String` and then `stop` would require such a path element. WDYT? I certainly need to think a little more about this.

patriknw replied:
I think it would be fine to use `ActorRef` and if it's not a direct child it would do nothing, same as if the childName doesn't exist

rkuhn replied:
convinced

patriknw commented on akka-typed/src/main/scala/akka/typed/ActorRef.scala:
```diff
@@ -0,0 +1,75 @@
+   * Implementation detail. The underlying untyped [[akka.actor.ActorRef]]
+   * of this typed ActorRef.
+   */
+  def ref: akka.actor.ActorRef
```
`untypedRef` ?

rkuhn replied:
yes, and probably also `private[akka]`

patriknw replied:
ah, even better, I thought there was some integration reason

patriknw commented on akka-typed/src/main/scala/akka/typed/patterns/Receptionist.scala:
```diff
@@ -0,0 +1,109 @@
+   * registrations can be made for the same key. Unregistration is implied by
+   * the end of the referenced Actor’s lifecycle.
+   */
+  case class Register[T](key: ServiceKey[T], address: ActorRef[T])(val replyTo: ActorRef[Registered[T]]) extends Command
```
final

patriknw commented on akka-typed/src/main/scala/akka/typed/patterns/Receiver.scala:
```diff
@@ -0,0 +1,122 @@
+  /**
+   * Retrieve one message from the Receiver, waiting at most for the given duration.
+   */
+  case class GetOne[T](timeout: FiniteDuration)(val replyTo: ActorRef[GetOneResult[T]]) extends Command[T]
```
final

patriknw commented:
I have reviewed a second time (except for tests). I'm comfortable with merging after the final touch.
Awesome work, @rkuhn !

rkuhn committed:
add akka-typed project with generic ActorRef

This is the first step towards more type-safety in Actor interactions,
comprising:

* generic ActorRef[T] that only accepts T messages
* generic ActorSystem[T] extends ActorRef[T] (sending to the guardian,
  whose Props[T] are provided for ActorSystem construction)
* removed the Actor trait: everything in there has been made into
  messages and signals
* new Behavior[T] abstraction that consumes messages (of type T) or
  Signals (lifecycle hooks, Terminated, ReceiveTimeout, Failed),
  producing the next Behavior[T] as the result each time
* the ask pattern is provided and yields properly typed Futures
* variants of ActorContext are provided for synchronous testing of
  Behaviors

All of this is implemented without touching code outside akka-typed
(apart from making guardianProps configurable), creating wrapper objects
around ActorRef, ActorContext, ActorSystem, Props and providing an Actor
implementation that just runs a Behavior.

viktorklang commented:
@patriknw Agreed, @rkuhn has managed to take Akka to the next level!

jboner commented on akka-docs/rst/scala/typed.rst:
```diff
@@ -0,0 +1,321 @@
+A Little Bit of Theory
+======================
+
+The `Actor Model`_ as defined by Hewitt, Bishop and Steiger in 1973 is a
```
I would add a link to the paper.

rkuhn replied:
You could open a PR :-)

jboner commented on akka-docs/rst/scala/typed.rst:
```diff
@@ -0,0 +1,321 @@
+reply fulfills a Promise to which we get back the corresponding Future.
+
+Note that the :class:`Future` that is returned by the “ask” operation is
+properly typed already, no type checks or casts needed. This is possible due to
```
I love this. Types all the way.

jboner commented on akka-docs/rst/scala/typed.rst:
```diff
@@ -0,0 +1,321 @@
+the feature that makes its type signatures and macros too complex to be viable.
+The solution chosen in Akka Typed is to explicitly include the properly typed
+reply-to address in the message, which both burdens the user with this task but
+also places this aspect of protocol design where it belongs.
```
Also removes the need for an Envelope, which is hard on the GC (at least long term, if we can remove the legacy Actor)

jboner commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,246 @@
+ *  - create a finite number of Actors
+ *  - designate the behavior for the next message
+ *
+ * In Akka the first capability is accessed by using the `!` or `tell` method
```
Love this. Very simple and clear.

jboner commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -0,0 +1,246 @@
+  /**
+   * Create a child Actor from the given [[Props]] and with the given name.
+   */
+  def spawn[U](props: Props[U], name: String): ActorRef[U]
```
Did we switch to 'spawn' just to be backward compatible and not clash? Else, why the change? 'actorOf' is pretty well established. 'spawn' sounds like Erlang-envy. Do we need to support untyped Actors here? For interop?

rkuhn replied:
At the time I couldn’t make type inference and overloading work together, and I have not looked back.

jboner commented on akka-typed/src/main/scala/akka/typed/ActorSystem.scala:
```diff
@@ -0,0 +1,144 @@
+
+/**
+ * An ActorSystem is home to a hierarchy of Actors. It is created using
+ * [[ActorSystem$.apply[T]*]] from a [[Props]] object that describes the root
```
Love this. Finally getting this straight. Turtles all the way up.

jboner commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,445 @@
+   * This behavior type is most useful for leaf actors that do not create child
+   * actors themselves.
+   */
+  final case class Total[T](behavior: T ⇒ Behavior[T]) extends Behavior[T] {
```
Will it not be hard for the user to know when to use which behavior type? My head is already spinning. Are there some simple guideline rules? Can we provide good defaults?

rkuhn replied:
The basic rule is: use whatever works. Using the “wrong” one will be very obvious, because it does not allow you to do what you want to do. Using an overly powerful version will just make the code a little more verbose, so no harm done.

jboner commented on akka-typed/src/main/scala/akka/typed/ScalaDSL.scala:
```diff
@@ -0,0 +1,445 @@
+   * {{{
+   * final case class OtherMsg(msg: String)(implicit val replyTo: ActorRef[Reply])
+   *
+   * SelfAware[MyCommand] { implicit self =>
```
I really like these decorators.
